File: packages/sdk/src/manifest.test.ts
import { Ajv2020 } from "ajv/dist/2020.js";
import { compile, type JSONSchema } from "json-schema-to-typescript";
import { describe, expect, it } from "vitest";
import type { Manifest } from "./manifest.js";
import schema from "./manifest.schema.json" with { type: "json" };

// The example manifest in docs/architecture.md §6.1.
const example = {
  kit: "wow",
  version: "0.1.0",
  sdk: "^0.1.0",
  tool_prefix: "wow",
  root: {
    locate: [
      { path: "{PROGRAM_FILES_X86}/World of Warcraft" },
      { path: "/Applications/World of Warcraft" },
      { prompt: "Select your World of Warcraft folder" },
    ],
    verify: "_*_",
  },
  adapter: {
    install: "_*_/Interface/AddOns/OgreMCP",
    process: ["Wow*.exe", "World of Warcraft*"],
  },
  sources: [
    {
      id: "savedvariables",
      type: "file",
      format: "text",
      path: "_*_/WTF/Account/*/SavedVariables/OgreMCP.lua",
      trigger: "on_change",
    },
  ],
  flavors: {
    classic_era: {
      status: "supported",
      search: ["https://www.wowhead.com/classic/", "https://warcraft.wiki.gg/"],
      mixed: ["https://warcraft.wiki.gg/"],
    },
    forever: { status: "experimental", search: [] },
  },
} satisfies Manifest;

const source = example.sources[0];

const validate = new Ajv2020({ allErrors: true, strict: true }).compile(schema);

describe("manifest schema", () => {
  it("accepts the §6.1 example", () => {
    expect(validate(example), JSON.stringify(validate.errors)).toBe(true);
  });

  it.each<[string, unknown]>([
    [
      "a path variable other than {PROGRAM_FILES_X86} and {HOME}",
      { ...example, root: { ...example.root, locate: [{ path: "{APPDATA}/World of Warcraft" }] } },
    ],
    [
      "a variable in a path relative to root",
      { ...example, sources: [{ ...source, path: "{HOME}/WTF/OgreMCP.lua" }] },
    ],
    ["an uppercase tool_prefix", { ...example, tool_prefix: "WoW" }],
    ["a tool_prefix with a dot", { ...example, tool_prefix: "wow.era" }],
    ["an unknown source type", { ...example, sources: [{ ...source, type: "log_tail" }] }],
    ["a kit ID that is not snake_case", { ...example, kit: "../x" }],
    ["a version that is not semver", { ...example, version: "0.1" }],
    ["the reserved flavor key unknown", { ...example, flavors: { unknown: example.flavors.forever } }],
    ...["http://www.wowhead.com/classic/", "https://www.wowhead.com/classic", "https://warcraft.wiki.gg", "https://warcraft.wiki.gg/?x=/"].map(
      (prefix): [string, unknown] => [
        `the search prefix ${prefix}, which is not https or does not end in /`,
        { ...example, flavors: { ...example.flavors, classic_era: { status: "supported", search: [prefix] } } },
      ],
    ),
  ])("rejects %s", (_case, manifest) => {
    expect(validate(manifest)).toBe(false);
  });

  it.each(["/", "\\Games", "C:\\Windows", "../../..", "_*_/../../.ssh/id_ed25519", "_*_\\..\\x"])(
    "rejects the path %s, which leaves root",
    (path) => {
      expect(validate({ ...example, adapter: { ...example.adapter, install: path } })).toBe(false);
    },
  );

  it.each(["Wow?.exe", "Wow[TB].exe", "bin/Wow.exe", "bin\\Wow.exe"])("rejects the process glob %s", (glob) => {
    expect(validate({ ...example, adapter: { ...example.adapter, process: [glob] } })).toBe(false);
  });

  // manifest.ts is generated from the schema, so the two cannot drift.
  it("matches the generated TS types in manifest.ts", async () => {
    const types = await compile(schema as JSONSchema, "Manifest", {
      bannerComment: [
        "// Generated from manifest.schema.json by manifest.test.ts. Do not edit.",
        "// After a schema change, regenerate with: pnpm --filter @ogremcp/sdk test -u",
      ].join("\n"),
      ignoreMinAndMaxItems: true,
    });
    await expect(types).toMatchFileSnapshot("./manifest.ts");
  });
});

File: bridge/internal/manifest/manifest_test.go
package manifest

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// wow parses the WoW kit's manifest. The test reads it from the repo; the
// bridge itself builds from nothing outside bridge/ (§5).
func wow(t *testing.T) *Manifest {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "..", "..", "kits", "wow", "manifest.json"))
	if err != nil {
		t.Fatal(err)
	}
	m, err := Parse(raw)
	if err != nil {
		t.Fatal(err)
	}
	return m
}

func TestParseWoWManifest(t *testing.T) {
	m := wow(t)
	if m.Kit != "wow" || m.ToolPrefix != "wow" {
		t.Errorf("kit %q, tool_prefix %q", m.Kit, m.ToolPrefix)
	}
	locate := m.Root.Locate
	if len(locate) < 2 || locate[0].Path != "{PROGRAM_FILES_X86}/World of Warcraft" || locate[len(locate)-1].Prompt == "" {
		t.Errorf("root.locate = %+v", locate)
	}
	if m.Root.Verify != "_*_" {
		t.Errorf("root.verify = %q", m.Root.Verify)
	}
	if m.Adapter == nil || m.Adapter.Install != "_*_/Interface/AddOns/OpenGamerMCP" || len(m.Adapter.Process) == 0 {
		t.Errorf("adapter = %+v", m.Adapter)
	}
	if len(m.Sources) != 1 || m.Sources[0].ID != "savedvariables" || m.Sources[0].Path != "_*_/WTF/Account/*/SavedVariables/OpenGamerMCP.lua" {
		t.Errorf("sources = %+v", m.Sources)
	}
	if m.Flavors["classic_era"].Status != "supported" {
		t.Errorf("flavors = %+v", m.Flavors)
	}
}

func TestParseRejectsLocateEntryShape(t *testing.T) {
	for _, entry := range []string{`{"path": "/a", "prompt": "b"}`, `{"steam": "123"}`} {
		raw := `{"kit": "wow", "root": {"locate": [` + entry + `], "verify": "x"}}`
		if _, err := Parse([]byte(raw)); err == nil || !strings.Contains(err.Error(), "exactly one of path and prompt") {
			t.Errorf("%s: %v", entry, err)
		}
	}
}

func TestValidateRejects(t *testing.T) {
	for _, tc := range []struct {
		name   string
		change func(m *Manifest)
		want   string
	}{
		{"absolute verify", func(m *Manifest) { m.Root.Verify = "/etc" }, "absolute"},
		{"backslash-rooted install", func(m *Manifest) { m.Adapter.Install = `\Windows\x` }, "absolute"},
		{"drive letter", func(m *Manifest) { m.Sources[0].Path = "C:/Windows/x" }, `":"`},
		{"dot-dot", func(m *Manifest) { m.Adapter.Install = "_*_/../../x" }, `".."`},
		{"dot-dot with backslashes", func(m *Manifest) { m.Sources[0].Path = `_*_\..\x` }, `".."`},
		{"dot-dot with a trailing space", func(m *Manifest) { m.Sources[0].Path = "_*_/.. /x" }, `".. "`},
		{"dot-dot with a trailing dot", func(m *Manifest) { m.Root.Verify = "..." }, `"..."`},
		{"names nothing", func(m *Manifest) { m.Root.Verify = "./" }, "names nothing"},
		{"variable in a relative path", func(m *Manifest) { m.Root.Verify = "{HOME}/x" }, "variable"},
		{"variable inside a locate path", func(m *Manifest) { m.Root.Locate[1].Path = "/x/{HOME}" }, "only start"},
		{"variable without a separator", func(m *Manifest) { m.Root.Locate[1].Path = "{HOME}x" }, "separator"},
		{"both path and prompt", func(m *Manifest) { m.Root.Locate[1].Prompt = "x" }, "exactly one"},
		{"duplicate source id", func(m *Manifest) { m.Sources = append(m.Sources, m.Sources[0]) }, "earlier source"},
		{"unknown source type", func(m *Manifest) { m.Sources[0].Type = "log_tail" }, "not file"},
		{"kit name", func(m *Manifest) { m.Kit = "../x" }, "snake_case"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			m := wow(t)
			tc.change(m)
			err := m.Validate()
			if err == nil || !strings.Contains(err.Error(), tc.want) {
				t.Errorf("Validate() = %v, want an error with %s", err, tc.want)
			}
		})
	}
}
